addons/web/controllers/main.py

```diff
@@ -702,8 +702,9 @@ def backup(self, master_pwd, name, backup_format = 'zip'):
     @http.route('/web/database/restore', type='http', auth="none", methods=['POST'], csrf=False)
     def restore(self, master_pwd, backup_file, name, copy=False):
         try:
-            data = base64.b64encode(backup_file.read())
-            dispatch_rpc('db', 'restore', [master_pwd, name, data, str2bool(copy)])
+            data = ''
+            for chunk in iter(lambda: backup_file.read(8190), b''):
+                data += base64.b64encode(chunk)            dispatch_rpc('db', 'restore', [master_pwd, name, data, str2bool(copy)])
```

Comment:
I think this is broken.

Also data could probably be a bytearray (explicitly mutable bytes) rather than rely on CPython optimisations.